src/backend/distributed/transaction/worker_transaction.c

```diff
@@ -0,0 +1,515 @@
+	if (workerConnection == NULL)
+	{
+		ereport(ERROR, (errmsg("could not open connection to %s:%d as %s",
+							   nodeName, nodePort, nodeUser)));
+	}
+
+	/* start the transaction on the worker node */
+	queryResult = PQexec(workerConnection, "BEGIN");
+	if (PQresultStatus(queryResult) != PGRES_COMMAND_OK)
+	{
+		ReraiseRemoteError(workerConnection, queryResult);
+	}
+
+	PQclear(queryResult);
+
+	/* iterate over the commands and execute them in the same connection */
+	foreach(commandCell, commandList)
+	{
+		char *commandString = lfirst(commandCell);
+		ExecStatusType resultStatus = PGRES_EMPTY_QUERY;
+
+		queryResult = PQexec(workerConnection, commandString);
+		resultStatus = PQresultStatus(queryResult);
+		if (!(resultStatus == PGRES_SINGLE_TUPLE || resultStatus == PGRES_TUPLES_OK ||
+			  resultStatus == PGRES_COMMAND_OK))
+		{
+			ReraiseRemoteError(workerConnection, queryResult);
+		}
+
+		PQclear(queryResult);
+	}
+
+	/* commit the transaction on the worker node */
+	queryResult = PQexec(workerConnection, "COMMIT");
+	if (PQresultStatus(queryResult) != PGRES_COMMAND_OK)
+	{
+		ReraiseRemoteError(workerConnection, queryResult);
+	}
+
+	PQclear(queryResult);
+
+	/* clear NULL result */
+	PQgetResult(workerConnection);
+
+	/* we no longer need this connection */
+	PQfinish(workerConnection);
+}
+
+
+/*
+ * IsWorkerTransactionActive returns true if there exists any on going
+ * worker transactions.
+ */
+bool
+IsWorkerTransactionActive(void)
+{
+	bool isWorkerTransactionActive = false;
+
+	if (workerConnectionList != NIL)
+	{
+		isWorkerTransactionActive = true;
+	}
+
+	return isWorkerTransactionActive;
+}
+
+
+/*
+ * EnableXactCallback registers the CompleteWorkerTransactions function as the callback
+ * of the worker transactions.
+ */
+static void
+EnableXactCallback(void)
+{
+	if (!isXactCallbackRegistered)
+	{
+		RegisterXactCallback(CompleteWorkerTransactions, NULL);
+		isXactCallbackRegistered = true;
+	}
+}
+
+
+/*
+ * CompleteWorkerTransaction commits or aborts pending worker transactions
+ * when the local transaction commits or aborts.
+ */
+static void
+CompleteWorkerTransactions(XactEvent event, void *arg)
+{
+	if (workerConnectionList == NIL)
+	{
+		/* nothing to do */
+		return;
+	}
+	else if (event == XACT_EVENT_PRE_COMMIT)
+	{
+		/*
+		 * Any failure here will cause local changes to be rolled back,
+		 * and may leave a prepared transaction on the remote node.
+		 */
+
+		PrepareRemoteTransactions(workerConnectionList);
+
+		/*
+		 * We are now ready to commit the local transaction, followed
+		 * by the remote transaction. As a final step, write commit
+		 * records to a table. If there is a last-minute crash
+		 * on the local machine, then the absence of these records
+		 * will indicate that the remote transactions should be rolled
+		 * back. Otherwise, the presence of these records indicates
+		 * that the remote transactions should be committed.
+		 */
+
+		LogPreparedTransactions(workerConnectionList);
```

Comment:
For MX, we were kind of free to make 2PC mandatory given. In this case, would it make sense to use the multi_shard_commit_protocol setting to avoid problems with max_prepared_transactions not being set on workers?